File: split-square/validate.js
// Is this a valid split square?

function validate(square) {
    if (square === 0 || square === 1) {
        return true;
    }

    if (Array.isArray(square) && square.length === 4) {
        return square.every(validate);
    }
    return false;
}
// Simple Square
console.log(validate(0)); // true

// A split square of four simple filled squares
console.log(validate([1, 1, 1, 1])) // true

// can nest split and simple squares;
console.log(validate([1, 0, [1, [0, 0, 0, 0], 1, [1, 1, 1, 1]], 1])); // true
console.log(validate([1,
    [1, 0, 1, [0, [0, 0, 0, 0], 1, 1]],
    [1, 0, 1, 0],
    1])); // true

// Simple square must be either empty (0) or filled (1)
console.log(validate(2)); // false

// Split squares must contain exactly four parts:
console.log(validate([1, 1, 1, 1, 1])); // false
console.log(validate([1, 0, [1, [0, 0, 0, 0, 1], 1, [1, 1, 1, 1]], 1])); // false
console.log(validate([1, [1, 0, 1, [0, [0, 0, 0], 1, 1]], [1, 0, 1, 0], 1])); // false

File: split-square/validate.py
"""Validate that a given square is valid.

Checks that this is either a simple square (``0`` or ``1``), or
a split square (a list of 4 items, each being a simple square or
a split square).

A simple square is valid::

    >>> validate(0)
    True

A split square of four simple filled squares is valid::

    >>> validate([1, 1, 1, 1])
    True

We can nest split and simple squares::

    >>> validate([1, 0, [1, [0, 0, 0, 0], 1, [1, 1, 1, 1]], 1])
    True

    >>> validate([1, [1, 0, 1, [0, [0, 0, 0, 0], 1, 1]], [1, 0, 1, 0], 1])
    True

Simple squares must be either 0 (empty) or 1 (filled)::

    >>> validate(2)
    False

Split squares must contain exactly four parts::

    >>> validate([1, 1, 1, 1, 1])
    False

    >>> validate([1, 0, [1, [0, 0, 0, 0, 1], 1, [1, 1, 1, 1]], 1])
    False

    >>> validate([1, [1, 0, 1, [0, [0, 0, 0], 1, 1]], [1, 0, 1, 0], 1])
    False

"""


def validate(square):
    """Validate that a given square is valid.."""

    # Base case - simple square if either 0 or 1
    if type(square) == int:
        return square == 0 or square == 1
    
    # Base case - needs to be a list of 4
    if type(square) != list or len(square) !=4:
        return False
    
    # Recurse into different quadrants
    for q in square:
        if not validate(q):
            return False
        
    return True


if __name__ == "__main__":
    import doctest
    if doctest.testmod().failed == 0:
        print "\n*** ALL TESTS PASS; THAT'S SUPER-VALID WORK!\n"
